### DISCUSSION TITLE
BM25 algorithm

### DISCUSSION THREAD

**Discussion by sorobedio** [seominjoon/kaist-ai605#106] [Q&A] [answered]
If i understand the BM25 as defined in the lecture note does not use cosine similarity it just outputs a scalar value for each document.
Right?

**Comment by seominjoon** [comment, marked as answer]
Are you asking if it is cosine similarity or inner product?

**Reply by sorobedio** [reply]
I meant in the given formula we sum up all terms(tokens) BM25 score per documents to create the documents scores which leads to a vector with length  being the number of documents rather than a matrix. Maybe i misread the formula. i think that the matrix form is much better with cosine similarity.

**Reply by seominjoon** [reply]
I am sorry, I don't think I understood your question. Perhaps an example will help.

**Reply by yuneg11** [reply]
I'm sorry to interrupt.
To implement `BM25` class, could we make it subclass `TFIDF` instead of `BagOfWords`?
(And also `GloVeAverage` for `GloVeAverageFaiss`?)

**Reply by seominjoon** [reply]
@yuneg11 Yes, you can subclass TFIDF and GloVeAverage instead if you would like. In general, when I said 'subclass', what I meant is that `isinstanceof(A, B)` should be true. It doesn't have to be the direct subclass though. Subclassing TFIDF is still also subclassing BagOfWords.

**Comment by sorobedio** [comment]
The BM25 in the lecture note is the same as the one presented on this page [https://en.wikipedia.org/wiki/Okapi_BM25]
it can be seen that the sum is over all term in the document. And if we sum that way it will result in a scalar value per document. but if we do not sum up we will have something similar to tf-idf

**Reply by seominjoon** [reply]
I see what you mean now. The easiest way to make this into an inner product is by defining the query vector with 1's only, and performing all necessary computations on the document side.

**Reply by emjay73** [reply]
Do we have to make this into an inner product? instead of directly computing the BM25 score?

**Reply by seominjoon** [reply]
For sparse vectors, you will have to compute the inner product yourself with a for loop anyways (you can't use PyTorch's tensor operation because the spare vectors are too large), so at the end, they are the same.
